File: go/mechanisms/evm/exact/facilitator/erc20approval.go
package facilitator

import (
	"encoding/hex"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"

	"github.com/coinbase/x402/go/extensions/erc20approvalgassponsor"
	"github.com/coinbase/x402/go/mechanisms/evm"
)

// approveSelector is the 4-byte function selector for ERC-20 approve(address,uint256).
// keccak256("approve(address,uint256)") = 0x095ea7b3...
var approveSelector = []byte{0x09, 0x5e, 0xa7, 0xb3}

// ValidateErc20ApprovalForPayment validates the ERC-20 approval extension data.
// Returns an empty string if valid, or an error reason string on failure.
func ValidateErc20ApprovalForPayment(info *erc20approvalgassponsor.Info, payer, tokenAddress string) string {
	// Step 1: Validate info format
	if !erc20approvalgassponsor.ValidateInfo(info) {
		return ErrErc20ApprovalInvalidFormat
	}

	// Step 2: Verify from matches payer
	if !strings.EqualFold(info.From, payer) {
		return ErrErc20ApprovalFromMismatch
	}

	// Step 3: Verify asset matches token
	if !strings.EqualFold(info.Asset, tokenAddress) {
		return ErrErc20ApprovalAssetMismatch
	}

	// Step 4: Verify spender is Permit2
	if !strings.EqualFold(info.Spender, evm.PERMIT2Address) {
		return ErrErc20ApprovalWrongSpender
	}

	// Step 5: Decode and parse the RLP transaction
	txHex := strings.TrimPrefix(info.SignedTransaction, "0x")
	txBytes, err := hex.DecodeString(txHex)
	if err != nil {
		return ErrErc20ApprovalTxParseFailed
	}

	tx := new(types.Transaction)
	if err := tx.UnmarshalBinary(txBytes); err != nil {
		return ErrErc20ApprovalTxParseFailed
	}

	// Step 6: Verify transaction target is the token contract
	if tx.To() == nil || !strings.EqualFold(tx.To().Hex(), tokenAddress) {
		return ErrErc20ApprovalWrongTarget
	}

	// Step 7: Verify the approve(address,uint256) selector
	data := tx.Data()
	if len(data) < 4 {
		return ErrErc20ApprovalWrongSelector
	}
	for i, b := range approveSelector {
		if data[i] != b {
			return ErrErc20ApprovalWrongSelector
		}
	}

	// Step 8: Decode spender from calldata (first 32-byte slot after selector, last 20 bytes)
	if len(data) < 36 {
		return ErrErc20ApprovalWrongCalldata
	}
	calldataSpender := common.BytesToAddress(data[4:36]) // last 20 bytes of first 32-byte param slot
	if !strings.EqualFold(calldataSpender.Hex(), evm.PERMIT2Address) {
		return ErrErc20ApprovalWrongCalldata
	}

	// Step 9: Verify the transaction was signed by payer
	chainID := tx.ChainId()
	signer := types.LatestSignerForChainID(chainID)
	from, err := types.Sender(signer, tx)
	if err != nil {
		return ErrErc20ApprovalInvalidSig
	}
	if !strings.EqualFold(from.Hex(), payer) {
		return ErrErc20ApprovalSignerMismatch
	}

	return ""
}

File: typescript/packages/mechanisms/evm/src/exact/facilitator/erc20approval.ts
import { getAddress, parseTransaction, decodeFunctionData, recoverTransactionAddress } from "viem";
import {
  validateErc20ApprovalGasSponsoringInfo,
  type Erc20ApprovalGasSponsoringInfo,
} from "@x402/extensions";
import { PERMIT2_ADDRESS } from "../../constants";

/** ERC-20 approve ABI for decoding approve(address,uint256) calldata */
const erc20ApproveAbi = [
  {
    type: "function",
    name: "approve",
    inputs: [
      { name: "spender", type: "address" },
      { name: "amount", type: "uint256" },
    ],
    outputs: [{ type: "bool" }],
    stateMutability: "nonpayable",
  },
] as const;

/** The approve(address,uint256) function selector */
const APPROVE_SELECTOR = "0x095ea7b3";

/**
 * Validates ERC-20 approval extension data for a Permit2 payment.
 *
 * Performs comprehensive validation:
 * - Format validation via validateErc20ApprovalGasSponsoringInfo
 * - `from` matches payer
 * - `asset` matches token
 * - `spender` is PERMIT2_ADDRESS
 * - Transaction `to` matches token address
 * - Transaction calldata is a valid approve(PERMIT2_ADDRESS, ...) call
 * - Recovered transaction signer matches `from`
 *
 * @param info - The ERC-20 approval gas sponsoring info
 * @param payer - The expected payer address
 * @param tokenAddress - The expected token address
 * @returns Validation result
 */
export async function validateErc20ApprovalForPayment(
  info: Erc20ApprovalGasSponsoringInfo,
  payer: `0x${string}`,
  tokenAddress: `0x${string}`,
): Promise<{ isValid: boolean; invalidReason?: string }> {
  // Validate format
  if (!validateErc20ApprovalGasSponsoringInfo(info)) {
    return { isValid: false, invalidReason: "invalid_erc20_approval_extension_format" };
  }

  // Verify from matches payer
  if (getAddress(info.from as `0x${string}`) !== getAddress(payer)) {
    return { isValid: false, invalidReason: "erc20_approval_from_mismatch" };
  }

  // Verify asset matches token
  if (getAddress(info.asset as `0x${string}`) !== tokenAddress) {
    return { isValid: false, invalidReason: "erc20_approval_asset_mismatch" };
  }

  // Verify spender field in info is Permit2
  if (getAddress(info.spender as `0x${string}`) !== getAddress(PERMIT2_ADDRESS)) {
    return { isValid: false, invalidReason: "erc20_approval_spender_not_permit2" };
  }

  // Parse and validate the signed transaction
  try {
    const serializedTx = info.signedTransaction as `0x${string}`;
    const tx = parseTransaction(serializedTx);

    // Verify the transaction targets the token contract
    if (!tx.to || getAddress(tx.to) !== tokenAddress) {
      return { isValid: false, invalidReason: "erc20_approval_tx_wrong_target" };
    }

    // Verify the calldata is an approve() call
    const data = tx.data ?? "0x";
    if (!data.startsWith(APPROVE_SELECTOR)) {
      return { isValid: false, invalidReason: "erc20_approval_tx_wrong_selector" };
    }

    // Decode the calldata to verify the spender argument is PERMIT2_ADDRESS
    try {
      const decoded = decodeFunctionData({
        abi: erc20ApproveAbi,
        data: data as `0x${string}`,
      });
      const calldataSpender = getAddress(decoded.args[0] as `0x${string}`);
      if (calldataSpender !== getAddress(PERMIT2_ADDRESS)) {
        return { isValid: false, invalidReason: "erc20_approval_tx_wrong_spender" };
      }
    } catch {
      return { isValid: false, invalidReason: "erc20_approval_tx_invalid_calldata" };
    }

    // Recover the signer of the transaction and verify it matches from
    try {
      const recoveredAddress = await recoverTransactionAddress({
        serializedTransaction: serializedTx,
      });
      if (getAddress(recoveredAddress) !== getAddress(payer)) {
        return { isValid: false, invalidReason: "erc20_approval_tx_signer_mismatch" };
      }
    } catch {
      return { isValid: false, invalidReason: "erc20_approval_tx_invalid_signature" };
    }
  } catch {
    return { isValid: false, invalidReason: "erc20_approval_tx_parse_failed" };
  }

  return { isValid: true };
}
